runtime/onert/backend/cpu/TensorBuilder.cc

```diff
@@ -94,20 +94,19 @@ void TensorBuilder::allocate()
 
 std::shared_ptr<ITensor> TensorBuilder::tensorAt(const ir::OperandIndex &ind)
 {
-  auto found = _tensor_reg->find(ind);
-  if (found == _tensor_reg->end())
-    return nullptr;
+  return _tensor_reg->getITensor(ind);
+}
 
-  return found->second;
+std::shared_ptr<IPortableTensor> TensorBuilder::portableAt(const ir::OperandIndex &ind)
```

Comment:
When is this method used?

Reply:
Oops, I missed explanation. This will be the next PR - https://github.com/wateret/ONE/commit/a1dbab5bec7e035b7a556cc40dbf3cc24c16ccf8 .

All calls `_tensor_builder->at()` will be replaced to `_tensor_builder->portableAt()` in CPU kernel generator and all `ops` classes in cpu backend is going to have `IPortableTensor` instead of `cpu_common::Tensor`.

Reply:
Could you add an assertion that checks portable tensor's layout in this method like `assert(_tensor_reg->getPortableTensor(ind)->layout() == ir::Layout::NHWC);`?

Reply:
Good point. Well, I have been thinking we can handle Layouts since just being portable tensors does not mean that all those tensors are reusable - we also have layout info in `PermuteFactor`. So it will skip inserting Permute operation only if layouts match.

---

For more details,

I'm going to remove Source/Sink when starting execution. Model Inputs are going to be treated as tensors in `controlflow`. That means Permute op will do the job of Source/Sink. This makes it `PermutationInsertionPass` to do everything so we can optimize it further.

For now `PermutationInsertionPass` always makes Permute operation since it cannot reuse tensors from other backends. But with this work, it can reuse `IPortableTensor` tensors so can skip inserting `Permute`s, **only if these conditions are met**.

- The tensor that we are processing is `IPortableTensor` the backend that uses that tensor can use portable tensors.
- Layouts must match (We have this in `PermuteFactor`)

These conditions will be implemented in `PermutationInsertionPass`.

Plus, this will also make it possible that, for example, if we have another CPU backend(let's say its id `cpu2` with different kernels), and some operations use cpu2 backend, then it will not insert any `Pemute` operations.

Reply:
With a discussion with @ragmani , I will add an assertion to CPU backend!